homeassistant/components/nest/__init__.py

```diff
@@ -192,20 +194,17 @@ async def async_handle_event(self, event_message: EventMessage) -> None:
             self._hass.bus.async_fire(NEST_EVENT, message)
 
 
-async def async_migrate_entry(hass: HomeAssistant, config_entry: ConfigEntry) -> bool:
-    """Migrate old entry."""
+def _entry_compat(config_entry: ConfigEntry) -> None:
+    """Update a ConfigEntry to newest format for backwards compatibility."""
     if DATA_SDM not in config_entry.data:
-        # Legacy API
-        return True
+        return
 
-    if config_entry.version == 1:
-        _LOGGER.debug("Migrating from version %s", config_entry.version)
-        new: Dict[str, Any] = {**config_entry.data}
-        new["auth_implementation"] = f"{DOMAIN}.web"
-        config_entry.version = 2
-        hass.config_entries.async_update_entry(config_entry, data=new)
-        _LOGGER.info("Migration to version %s successful", config_entry.version)
-    return True
+    if config_entry.data.get("auth_implementation") == DOMAIN:
+        # SDM API previously only supported a single auth implementation type
+        # so update on demand to new name for the original impl.
+        new = {**config_entry.data}
+        new["auth_implementation"] = WEB_AUTH_DOMAIN
+        config_entry.data = MappingProxyType({**new})
```

Comment:
Please use `hass.config_entries.async_update_entry` to update the config entry.

Reply:
Ah! I assumed that `async_update_entry` was equivalent to doing a hard migration. Is doing an update here only preserved at runtime? I'd love to learn more about the differences between the migration path and doing an update.

Also curious about implications in testing here (e.g. should tests verify an existing configuration entry was modified)

Reply:
A hard migration is when we change the config entry version and/or the config entry data to make it backwards incompatible.

Will the old code work with the new name of the implementation in config entry data?

Reply:
Ah good point. So the case is:
- Install home assistant 2021.11
- Sign up and create a config entry
- Install home assistant 2021.12
- Update config entry
- Revert back to 2021.11

As written, this will not work.

So perhaps its not worth trying to update the existing auth implementation name and I should leave it alone.  I've gone ahead and reverted the domain for web auth, but left around the variable names (e.g. `WEB_AUTH_DOMAIN = DOMAIN`)

Thank you for the detailed review.

Reply:
Yeah, then we can remove the entry compat function. 👍

Reply:
Good catch, removed.